### PR TITLE
[examples/with-typescript-graphql] Add "createResolverContext" hook, etc.

### PR DESCRIPTION
**Pull request by piglovesyou**
I added a few features and updated dependencies.

**feat: Add resolver context**
* This allows developers to integrate session information to GraphQL resolvers
* Remove JSDoc, respect TypeScript defs

**feat: Add graphql-let/schema/loader**
* This enables HMR from modifying GraphQL schema

~~**Simplify "isSSR" logic**~~ Not included any more
* ~~Breaking change. The second argument of `withApollo` was confusing because there's `typeof window === 'undefined'` already. I think user don't have to specify if they use `getDataFromTree` per page.~~

**Update deps**
* Run `node-check-update -u`
* "graphql" keeps ^14 since other libs expect it
* Install apollo-link and apollo-cache because @apollo/react-common wants it as peer-deps

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| buildDuration | 10.4s | 10.5s | ⚠️ +65ms |
| nodeModulesSize | 47.7 MB | 47.7 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.25 kB | 6.25 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..5272.js gzip | 10.2 kB | 10.2 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.3 kB | 56.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.75 kB | 6.75 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| index.html gzip | 917 B | 917 B | ✓ |
| link.html gzip | 925 B | 925 B | ✓ |
| withRouter.html gzip | 914 B | 914 B | ✓ |
| Overall change | 2.76 kB | 2.76 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| buildDuration | 11.5s | 11.1s | -396ms |
| nodeModulesSize | 47.7 MB | 47.7 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.25 kB | 6.25 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..5272.js gzip | 10.2 kB | 10.2 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.3 kB | 56.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.75 kB | 6.75 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | piglovesyou/next.js update-wtg | Change |
| - | - | - | - |
| _error.js gzip | 233 kB | 233 kB | ⚠️ +142 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 956 B | 956 B | ✓ |
| index.js gzip | 233 kB | 233 kB | ⚠️ +21 B |
| link.js gzip | 243 kB | 243 kB | ⚠️ +256 B |
| routerDirect.js gzip | 241 kB | 241 kB | ⚠️ +17 B |
| withRouter.js gzip | 241 kB | 241 kB | ⚠️ +47 B |
| Overall change | 1.19 MB | 1.19 MB | ⚠️ +483 B |

</details>

</details>

**Review by lfades — CHANGES_REQUESTED** [review]
Please undo anything related to `Simplify "isSSR" logic`.

It was made like that intentionally, a page can have `{ ssr: false }` and then it could be statically optimized instead of being exported as a serverless function. Also the check `typeof window === 'undefined'` has to be kept inline (it gets replaced by a boolean) instead of using a variable so the code that's not required for the client side or server side build is removed by webpack.

**Comment by piglovesyou** [comment]
@itsfadnis Thank you for the reviewing, I didn't know those two. Let me fix them and rebase the branch.

File: examples/with-typescript-graphql/lib/with-apollo.tsx [hunk 1]
```diff
@@ -54,11 +73,25 @@ export default function withApollo(
...
+      let resolverContext: ResolverContext | undefined
+      // Keep the "isServer" check inline, so webpack removes the block
+      // for client-side bundle.
+      if (typeof window === 'undefined') {
+        if (!ctx.req || !ctx.res) throw new Error('never')
+        resolverContext = await createResolverContext({
+          req: ctx.req,
+          res: ctx.res,
+        })
+      }
```
Comment (lfades) [inline examples/with-typescript-graphql/lib/with-apollo.tsx:87]: What would be the use case for this?

File: examples/with-typescript-graphql/lib/with-apollo.tsx [hunk 2]
```diff
@@ -54,11 +73,25 @@ export default function withApollo(
...
   if (ssr || PageComponent.getInitialProps) {
     WithApollo.getInitialProps = async (ctx: WithApolloPageContext) => {
-      const { AppTree } = ctx
+      // Resolver context here is only set on server. For client-side,
+      // "/api/graphql" route creates and pass it to resolver functions.
+      let resolverContext: ResolverContext | undefined
+      // Keep the "isServer" check inline, so webpack removes the block
+      // for client-side bundle.
+      if (typeof window === 'undefined') {
+        if (!ctx.req || !ctx.res) throw new Error('never')
```
Comment (lfades) [inline examples/with-typescript-graphql/lib/with-apollo.tsx:82]:
```suggestion
```

File: examples/with-typescript-graphql/lib/with-apollo.tsx (hunk 1 above)
Comment (piglovesyou) [reply examples/with-typescript-graphql/lib/with-apollo.tsx:87]: Thank you for asking. It's necessary for SSR to pass a resolver context to resolver functions. For SSR, it:

0. Starts from `WithApollo.getInitialProps`
1. Creates a resolver context from `req` and `res`
2. Initializes an apollo client instance with [`SchemaLink`](https://github.com/zeit/next.js/pull/11681/files#diff-52c0088e3df4611902ee3b5f59a4a97dR196) with the resolver context
3. Executes `getDataFromTree`, which leads the apollo client to...
4. Pass the resolver context to [resolver functions as the 3rd argument](https://github.com/zeit/next.js/pull/11681/files#diff-682d12e5f9468d1190b78c5c58a34533R5)

To pass a resolver context to resolver functions, we have to take care of two paths.

1. One for an  HTTP request to the GraphQL server (it's in `pages/api/graphql.ts`)
1. Another for SSR, done by `SchemaLink` that allows us to resolve GraphQL query without calling HTTP request.

The above code covers the case no. 2.

As you know, both paths run on the server-side, so here we have to get off client-side execution.

File: examples/with-typescript-graphql/lib/with-apollo.tsx (hunk 2 above)
Comment (piglovesyou) [reply examples/with-typescript-graphql/lib/with-apollo.tsx:82]: Thanks, but how do you think we can avoid the TypeScript error?
Screen Shot 2020-04-09 at 9 34 54

**Comment by piglovesyou** [comment]
I'd like to follow the update of the dependencies. Let me please add a few more changes.

**Comment by piglovesyou** [comment]
@lfades I'm glad if you take a look at this again, thanks.

File: examples/with-typescript-graphql/lib/with-apollo.tsx (hunk 1 above)
Comment (lfades) [reply examples/with-typescript-graphql/lib/with-apollo.tsx:87]: @piglovesyou Thank you, it's very clear now! :relieved:

File: examples/with-typescript-graphql/lib/with-apollo.tsx (hunk 2 above)
Comment (lfades) [reply examples/with-typescript-graphql/lib/with-apollo.tsx:82]: With `ctx.req!` :+1: 

It will always be defined in this case.

**Review by lfades — APPROVED** [review]
@piglovesyou Thank you!

**Comment by piglovesyou** [comment]
Thank you for following up!

**Pull request merged**